doc/Configuration.md
# Configuration file

This file allows the user to customize the behavior of AeonVoice. It is read-only once at startup, so changes will take effect only after restarting the application which is using AeonVoice.

On Windows, the configuration file is named `AeonVoice.ini`. on other platforms it is named `AeonVoice.conf`.

## Path

### Linux

The path to the directory where the settings are stored is determined at compile time. It can be redefined by setting the value of the `sysconfdir` variable when executing `scons`. By default, the value is `$prefix/etc`, which, if the default value of the `prefix` variable is used, meaning `/usr/local/etc`.

The full path to the configuration file is `$sysconfdir/AeonVoice/AeonVoice.conf`.

### Windows

#### SAPI5

The path is `%APPDATA%\AeonVoice\AeonVoice.ini`.

## Format

AeonVoice uses the classic ini format. But as this format is not defined by a formal standard and implementations differ, here are some implementation details users should keep in mind.

* The file must be saved in UTF-8.
* Semicolons are used to indicate comments. Only the whole line can be commented out.
* Periods are used as decimal separators.
* Characters in values may be represented with their numeric codes. The format is similar to numeric character references in XML, but the initial ampersand sign is omitted. For example, the English letter A can be written as `#65;`.
* If an option takes only two boolean values, on or off,  use the following keywords to enable the corresponding behavior: `true`, `yes`, `on` or `1`, and to disable it, use `false`, `no`, `off` or `0`.

## General and specific settings

Some options can be applied to a specific installed voice.

### English-wide settings

The following format is used:

```ini
languages.<language>.<key>=<value>
```

AeonVoice supports English only. Use `English`, `en`, or `eng`.

Examples:

```ini
languages.english.default_rate=0.8
languages.eng.default_rate=0.8
```

### Voice-specific settings

The following format is used:

```ini
voices.<name>.<key>=<value>
```

Example:

```ini
voices.leena.enabled=no
```

## Available options

### Rate, pitch and volume

The settings in this group determine how AeonVoice interprets the values of rate, pitch and volume it receives from client programs.

The value `1` corresponds to the standard behavior of the voices.

For example, for an expression of the value of speech rate as a percentage, AeonVoice interprets 50% as the default speech rate (`default_rate`), and 100% is interpreted as the maximum speech rate (`max_rate`). The following table lists all the options in this group.

| Setting          | Description    | Default value | Minimum value | Maximum value |
| ---------------- | -------------- | ------------- | ------------- | ------------- |
| `default_rate`   | default rate   | 1             | `min_rate`    | `max_rate`    |
| `min_rate`       | minimum rate   | 0.5           | 0.2           | 1             |
| `max_rate`       | maximum rate   | 2             | 1             | 5             |
| `default_pitch`  | default pitch  | 1             | `min_pitch`   | `max_pitch`   |
| `min_pitch`      | minimum pitch  | 0.5           | 0.5           | 1             |
| `max_pitch`      | maximum pitch  | 2             | 1             | 2             |
| `default_volume` | default volume | 1             | `min_volume`  | `max_volume`  |
| `min_volume`     | minimum volume | 0.25          | 0.25          | 1             |
| `max_volume`     | maximum volume | 2             | 1             | 4             |

These settings can be applied to all voices, English as a whole, or an individual voice.

Examples:

```ini
default_volume=0.8
languages.english.default_rate=1.5
voices.toptygin.default_pitch=0.9
```

#### Using the Sonic library

AeonVoice can use [the Sonic library](https://github.com/waywardgeek/sonic) for changing speech rate. The native algorithm of speech rate modification in AeonVoice was improved in version 1.4.0, and this library is not included in the builds provided by the AeonVoice project. So the following setting is only supported in custom builds with Sonic enabled.

The `min_sonic_rate` setting specifies the minimum rate value starting from which Sonic will be used instead of the built-in algorithm. By default the built-in algorithm is always used on Android, and on other platforms Sonic is used to speed up speech, that is the synthesizer behaves as if the user wrote in the configuration file:

```ini
min_sonic_rate=1
```

### Voice profiles

AeonVoice currently synthesizes English only. A voice profile selects one
installed English voice for clients that support profile selection. The engine
also makes an individual profile available for each installed voice.

The `voice_profiles` setting is a comma-separated list of profile names. The
included configuration declares the `Toptygin` and `Leena` profiles:

```ini
voice_profiles=Toptygin,Leena
```

### Speech quality

The `quality` setting allows the user to choose one of the three available speech quality options.  The lower the quality, the better the performance: most importantly, the less time it will take for the synthesizer to start speaking. The available modes are described in the following table.

| Value      | Sampling rate (kHz) | Optimized response time |
| ---------- | ------------------- | ----------------------- |
| `max`      | 24                  | no                      |
| `standard` | 24                  | yes                     |
| `min`      | 16                  | yes                     |

The standard quality is used by default, that is the synthesizer behaves as if the user wrote in the configuration file:

```ini
quality=standard
```

### Punctuation

Despite the title, the settings in this group apply to other non-alphabetic characters as well, even though, strictly speaking, they can't be considered punctuation marks.

#### Punctuation mode

The `punctuation_mode` setting determines whether punctuation marks and other symbols will be spoken. The following modes are supported.

| Value  | Description                      |
| ------ | -------------------------------- |
| `none` | don't read any symbols (default) |
| `some` | selective reading                |
| `all`  | read all symbols                 |

For example, reading of all punctuation is enabled this way:

```ini
punctuation_mode=all
```

#### Selective reading of punctuation

The `punctuation_list` setting is used to specify which symbols to speak when the punctuation mode is set to `some`. For example:

```ini
punctuation_list=@$/\
```

#### Names of symbols

The built-in dictionary of punctuation marks and other symbols can't be called comprehensive.  But users can define the names of additional symbols in the user dictionary.

### Capital letters

The following settings determine whether and how the synthesizer will indicate capital letters when it reads individual characters.

#### Capital letter indication mode

This mode is specified by the `indicate_capitals` setting. The following values are supported.

| Value   | Description                   |
| ------- | ----------------------------- |
| `no`    | disabled (default)            |
| `word`  | say "capital" before a letter |
| `pitch` | change speech pitch           |
| `sound` | play a short sound            |

Example:

```ini
indicate_capitals=pitch
```

#### Changing pitch to indicate capital letters

If capital letters are indicated by a change in pitch, than the `cap_pitch_factor` setting specifies how much to change the pitch in this situation. This setting can be applied both to the synthesizer as a whole and to individual voices. The following example specifies a thirty percent increase:

```ini
cap_pitch_factor=1.3
```

### Disabling individual languages and voices

You can disable English or an individual voice. The following example disables English and Toptygin:

```ini
languages.english.enabled=false
voices.toptygin.enabled=false
```


doc/Configuration-file.md
# Configuration file

This file allows the user to customize the behavior of AeonVoice. It is read-only once at startup, so changes will take effect only after restarting the application which is using AeonVoice.

On Windows, the configuration file is named `AeonVoice.ini`. on other platforms it is named `AeonVoice.conf`.

## Path

### Linux

The path to the directory where the settings are stored is determined at compile time. It can be redefined by setting the value of the `sysconfdir` variable when executing `scons`. By default, the value is `$prefix/etc`, which, if the default value of the `prefix` variable is used, meaning `/usr/local/etc`.

The full path to the configuration file is `$sysconfdir/AeonVoice/AeonVoice.conf`.

### Windows

#### SAPI5

The path is `%APPDATA%\AeonVoice\AeonVoice.ini`.

## Format

AeonVoice uses the classic ini format. But as this format is not defined by a formal standard and implementations differ, here are some implementation details users should keep in mind.

* The file must be saved in UTF-8.
* Semicolons are used to indicate comments. Only the whole line can be commented out.
* Periods are used as decimal separators.
* Characters in values may be represented with their numeric codes. The format is similar to numeric character references in XML, but the initial ampersand sign is omitted. For example, the English letter A can be written as `#65;`.
* If an option takes only two boolean values, on or off,  use the following keywords to enable the corresponding behavior: `true`, `yes`, `on` or `1`, and to disable it, use `false`, `no`, `off` or `0`.

## General and specific settings

Some of the options allow specification that they must be applied only to a specific language or voice. And some of them are only defined for specific languages.

### Language-specific settings

The following format is used:

```ini
languages.<language>.<key>=<value>
```

The language can be indicated by:

* Its English name
* its two-letter code defined by ISO 639-1
* its three-letter code defined by ISO 639-3.

Examples:

```ini
languages.russian.default_rate=2
languages.eng.default_rate=0.8
```

### Voice-specific settings

The following format is used:

```ini
voices.<name>.<key>=<value>
```

Example:

```ini
voices.elena.enabled=no
```

## Available options

### Rate, pitch and volume

The settings in this group determine how AeonVoice interprets the values of rate, pitch and volume it receives from client programs.

The value `1` corresponds to the standard behavior of the voices.

For example, for an expression of the value of speech rate as a percentage, AeonVoice interprets 50% as the default speech rate (`default_rate`), and 100% is interpreted as the maximum speech rate (`max_rate`). The following table lists all the options in this group.

| Setting          | Description    | Default value | Minimum value | Maximum value |
| ---------------- | -------------- | ------------- | ------------- | ------------- |
| `default_rate`   | default rate   | 1             | `min_rate`    | `max_rate`    |
| `min_rate`       | minimum rate   | 0.5           | 0.2           | 1             |
| `max_rate`       | maximum rate   | 2             | 1             | 5             |
| `default_pitch`  | default pitch  | 1             | `min_pitch`   | `max_pitch`   |
| `min_pitch`      | minimum pitch  | 0.5           | 0.5           | 1             |
| `max_pitch`      | maximum pitch  | 2             | 1             | 2             |
| `default_volume` | default volume | 1             | `min_volume`  | `max_volume`  |
| `min_volume`     | minimum volume | 0.25          | 0.25          | 1             |
| `max_volume`     | maximum volume | 2             | 1             | 4             |

These settings can be applied both to all the voices and to individual languages or voices.

Examples:

```ini
default_volume=0.8
languages.english.default_rate=1.5
voices.alan.default_pitch=0.9
```

#### Using the Sonic library

AeonVoice can use [the Sonic library](https://github.com/waywardgeek/sonic) for changing speech rate. The native algorithm of speech rate modification in AeonVoice was improved in version 1.4.0, and this library is not included in the builds provided by the AeonVoice project. So the following setting is only supported in custom builds with Sonic enabled.

The `min_sonic_rate` setting specifies the minimum rate value starting from which Sonic will be used instead of the built-in algorithm. By default the built-in algorithm is always used on Android, and on other platforms Sonic is used to speed up speech, that is the synthesizer behaves as if the user wrote in the configuration file:

```ini
min_sonic_rate=1
```

### Language switching and voice profiles
 AeonVoice can analyze the input text and change the synthesis language automatically. This functionality is currently implemented only for those
languages whose alphabets share no letters.

#### Voice profiles

A voice profile is just a combination of voices selected by the user. The languages of the voices must differ. The first voice in the group is considered the primary voice and determines the so-called primary language.

A voice profile is written as a sequence of names of voices separated by the `+` character. The `voice_profiles` setting is used to define the list of voice profiles. Its value must be a comma-separated list of voice profiles.

Example:

```ini
voice_profiles=anna+slt,aleksandr+alan
```

#### Language switching

When one of the voice profiles is active, AeonVoice will try to use the voice which corresponds to the language of the text. The language may be set by the client program or detected automatically. If the language cannot be detected automatically, as is the case when reading individual numbers or names of punctuationmarks, the primary voice will be used, that is the first voice of the profile.

But if numbers or non-alphabetic characters appear in a sentence, and AeonVoice determines that the language of the sentence differs from the primary language, there are two possibilities: read them with the primary voice or read the whole sentence with one voice according to the language. By default the first alternative is chosen, but the user can change this behavior by disabling the `prefer_primary_language` setting:

```ini
prefer_primary_language=false
```

### Speech quality

The `quality` setting allows the user to choose one of the three available speech quality options.  The lower the quality, the better the performance: most importantly, the less time it will take for the synthesizer to start speaking. The available modes are described in the following table.

| Value      | Sampling rate (kHz) | Optimized response time |
| ---------- | ------------------- | ----------------------- |
| `max`      | 24                  | no                      |
| `standard` | 24                  | yes                     |
| `min`      | 16                  | yes                     |

The standard quality is used by default, that is the synthesizer behaves as if the user wrote in the configuration file:

```ini
quality=standard
```

### Punctuation

Despite the title, the settings in this group apply to other non-alphabetic characters as well, even though, strictly speaking, they can't be considered punctuation marks.

#### Punctuation mode

The `punctuation_mode` setting determines whether punctuation marks and other symbols will be spoken. The following modes are supported.

| Value  | Description                      |
| ------ | -------------------------------- |
| `none` | don't read any symbols (default) |
| `some` | selective reading                |
| `all`  | read all symbols                 |

For example, reading of all punctuation is enabled this way:

```ini
punctuation_mode=all
```

#### Selective reading of punctuation

The `punctuation_list` setting is used to specify which symbols to speak when the punctuation mode is set to `some`. For example:

```ini
punctuation_list=@$/\
```

#### Names of symbols

The built-in dictionary of punctuation marks and other symbols can't be called comprehensive.  But users can define the names of additional symbols in the user dictionary.

### Capital letters

The following settings determine whether and how the synthesizer will indicate capital letters when it reads individual characters.

#### Capital letter indication mode

This mode is specified by the `indicate_capitals` setting. The following values are supported.

| Value   | Description                   |
| ------- | ----------------------------- |
| `no`    | disabled (default)            |
| `word`  | say "capital" before a letter |
| `pitch` | change speech pitch           |
| `sound` | play a short sound            |

Example:

```ini
indicate_capitals=pitch
```

#### Changing pitch to indicate capital letters

If capital letters are indicated by a change in pitch, than the `cap_pitch_factor` setting specifies how much to change the pitch in this situation. This setting can be applied both to the synthesizer as a whole and to individual voices. The following example specifies a thirty percent increase:

```ini
cap_pitch_factor=1.3
```

### Stress marks

Stress indication in a text is supported only for Russian language. For example, to tell the synthesizer that the vowel after the `+` sign must receive stress, Add the following to the configuration file:

```ini
stress_marker=+
```

### Disabling individual languages and voices

You can disable some of the languages or voices. It may be useful if you don't use them and don't want your screen reader to show them. The following example disables all the English voices and Aleksandr::

```ini
languages.english.enabled=false
voices.aleksandr.enabled=false
```

### Pseudo-English mode

This mode is available only if the language doesn't use the Latin alphabet. When they read English words, the voices can try to follow the correct English pronunciation as much as possible. Though they can't use the correct English sounds, but the closest sounds in their native language.

To use this mode, you need to install the English language pack. The English voices don't need to be installed.

If the English language pack is installed, the pseudo-English mode is enabled by default. Using Russian as an example, you can disable it this way:

```ini
languages.russian.use_pseudo_english=no
```


src/core/params.cpp
/* SPDX-License-Identifier: GPL-3.0-or-later */

/* This program is free software: you can redistribute it and/or modify */
/* it under the terms of the GNU General Public License as published by */
/* the Free Software Foundation, either version 3 of the License, or */
/* (at your option) any later version. */

/* This program is distributed in the hope that it will be useful, */
/* but WITHOUT ANY WARRANTY; without even the implied warranty of */
/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the */
/* GNU General Public License for more details. */

/* You should have received a copy of the GNU General Public License */
/* along with this program.  If not, see <https://www.gnu.org/licenses/>. */

#include "core/params.hpp"

namespace
{
  const double MAX_MAX_VOLUME=4;
  const double MAX_MAX_RATE=5;
  const double NO_MIN_SONIC_RATE=MAX_MAX_RATE+1;
}

#ifndef MAX_VOLUME
#define MAX_VOLUME 2
#endif

#ifndef MAX_RATE
#define MAX_RATE 2
#endif

#ifndef MIN_SONIC_RATE
#define MIN_SONIC_RATE 1
#endif

#ifndef DEFAULT_PUNCTUATION_MODE
#define DEFAULT_PUNCTUATION_MODE AeonVoice_punctuation_none
#endif

namespace AeonVoice
{
  voice_params::voice_params():
    default_rate("default_rate",1,0.2,MAX_MAX_RATE),
    min_rate("min_rate",0.5,0.2,1),
    max_rate("max_rate",MAX_RATE,1,MAX_MAX_RATE),
    default_pitch("default_pitch",1,0.5,2),
    min_pitch("min_pitch",0.5,0.5,1),
    max_pitch("max_pitch",2,1,2),
    default_volume("default_volume",1,0.25,MAX_MAX_VOLUME),
    min_volume("min_volume",0.25,0.25,1),
    max_volume("max_volume",MAX_VOLUME,1,MAX_MAX_VOLUME),
    cap_pitch_factor("cap_pitch_factor",1.3,0.5,2),
    min_sonic_rate("min_sonic_rate",MIN_SONIC_RATE,0.2,NO_MIN_SONIC_RATE)
  {
  }

  void voice_params::default_to(const voice_params& other)
  {
    default_rate.default_to(other.default_rate);
    min_rate.default_to(other.min_rate);
    max_rate.default_to(other.max_rate);
    default_pitch.default_to(other.default_pitch);
    min_pitch.default_to(other.min_pitch);
    max_pitch.default_to(other.max_pitch);
    default_volume.default_to(other.default_volume);
    min_volume.default_to(other.min_volume);
    max_volume.default_to(other.max_volume);
    cap_pitch_factor.default_to(other.cap_pitch_factor);
    min_sonic_rate.default_to(other.min_sonic_rate);
  }

  void voice_params::register_self(config& cfg,const std::string& prefix)
  {
    cfg.register_setting(default_rate,prefix);
    cfg.register_setting(min_rate,prefix);
    cfg.register_setting(max_rate,prefix);
    cfg.register_setting(default_pitch,prefix);
    cfg.register_setting(min_pitch,prefix);
    cfg.register_setting(max_pitch,prefix);
    cfg.register_setting(default_volume,prefix);
    cfg.register_setting(min_volume,prefix);
    cfg.register_setting(max_volume,prefix);
    cfg.register_setting(cap_pitch_factor,prefix);
    cfg.register_setting(min_sonic_rate,prefix);
  }

  text_params::text_params():
    stress_marker("stress_marker",'+')
  {
  }

  void text_params::default_to(const text_params& other)
  {
    stress_marker.default_to(other.stress_marker);
  }

  void text_params::register_self(config& cfg,const std::string& prefix)
  {
    cfg.register_setting(stress_marker,prefix);
  }

  absolute_speech_params::absolute_speech_params():
    rate("rate",0,-1,1),
    pitch("pitch",0,-1,1),
    volume("volume",0,-1,1)
  {
  }

  void absolute_speech_params::default_to(const absolute_speech_params& other)
  {
    rate.default_to(other.rate);
    pitch.default_to(other.pitch);
    volume.default_to(other.volume);
  }

  relative_speech_params::relative_speech_params():
    rate("rate",1,0.1,10),
    pitch("pitch",1,0.5,2),
    volume("volume",1,0.25,2)
  {
  }

  void relative_speech_params::default_to(const relative_speech_params& other)
  {
    rate.default_to(other.rate);
    pitch.default_to(other.pitch);
    volume.default_to(other.volume);
  }

  verbosity_params::verbosity_params():
    punctuation_mode("punctuation_mode",DEFAULT_PUNCTUATION_MODE),
    punctuation_list("punctuation_list","+=<>~@#$%^&*|"),
    capitals_mode("indicate_capitals",AeonVoice_capitals_off)
  {
    punctuation_mode.define("none",AeonVoice_punctuation_none);
    punctuation_mode.define("some",AeonVoice_punctuation_some);
    punctuation_mode.define("all",AeonVoice_punctuation_all);
    capitals_mode.define("off",AeonVoice_capitals_off);
    capitals_mode.define("no",AeonVoice_capitals_off);
    capitals_mode.define("none",AeonVoice_capitals_off);
    capitals_mode.define("false",AeonVoice_capitals_off);
    capitals_mode.define("word",AeonVoice_capitals_word);
    capitals_mode.define("pitch",AeonVoice_capitals_pitch);
    capitals_mode.define("sound",AeonVoice_capitals_sound);
  }

  void verbosity_params::default_to(const verbosity_params& other)
  {
    punctuation_mode.default_to(other.punctuation_mode);
    punctuation_list.default_to(other.punctuation_list);
    capitals_mode.default_to(other.capitals_mode);
  }

  void verbosity_params::register_self(config& cfg,const std::string& prefix)
  {
    cfg.register_setting(punctuation_mode,prefix);
    cfg.register_setting(punctuation_list,prefix);
    cfg.register_setting(capitals_mode,prefix);
  }

  void stream_params::register_self(config& cfg)
  {
    cfg.register_setting(fixed_size);
    cfg.register_setting(view_size);
  }
}
